ligo/src/token/fa2.mligo

```diff
@@ -0,0 +1,176 @@
+// SPDX-FileCopyrightText: 2020 TQ Tezos
+// SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+// Corresponds to Token/FA2.hs module
+
+#include "../types.mligo"
+
+let frozen_token_id: nat = 0n
+let unfrozen_token_id: nat = 1n
+// let (|>) x f = f x
+
+// -----------------------------------------------------------------
+// Helper
+// -----------------------------------------------------------------
+
+let ensure_not_migrated (storage : storage): bool =
+  match storage.migration_status with
+      Not_in_migration -> true
+    | MigratingTo (p) -> true
+    | MigratedTo (p) -> (failwith ("MIGRATED") : bool)
+
+let check_sender (from_ , store : address * storage): bool =
+  if (sender = from_) then true
+  else
+    let key: operator = { owner = from_; operator = sender} in
+    if Big_map.mem key store.operators then
+      true
+    else
+      (failwith("FA2_NOT_OPERATOR"): bool)
+
+let debit_from (amt, from_, token_id : nat * address * token_id) (store : storage): storage =
+  let bal_o = Big_map.find_opt (from_, token_id) store.ledger in
+  match bal_o with
+    Some bal ->
+      let new_bal_int = bal - amt in
+      let updated_store =
+        match Michelson.is_nat (new_bal_int) with
+          Some new_bal ->
+            { store with
+              ledger = Big_map.update (from_, token_id) (Some new_bal) store.ledger
+            }
+        | None ->
+            // TODO: How to fail with some value
+            // (failwith ("FA2_INSUFFICIENT_BALANCE", (amt, bal)) : storage)
+            (failwith ("FA2_INSUFFICIENT_BALANCE") : storage)
```

Comment:
Seems like we can `failwith` something else beside a string by using something similar to this: https://github.com/tqtezos/stablecoin/blob/master/ligo/stablecoin/actions.ligo#L717-L726

Do you think it is worth doing or just leave it as `string` for now?

Reply:
I didn't think about that. I think it is worth doing, but let's not do this here.

I have specified in the specification that the errors are returned as `string` only too, so that needs to be updated as well.

I'll open a separate issue to add this different `failwith` and probably work on it myself in parallel.